### PR TITLE
[RLlib] SlateQ torch fixes and cleanups.

### PR DESCRIPTION
**Pull request by sven1977**
SlateQ torch fixes and cleanups.
* SlateQ torch now works exactly the same as its tf2 version **plus** the choice-loss term.
* Added CI learning test case for interest-evolution env.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/slateq/slateq.py [hunk 1]
```diff
@@ -113,21 +106,17 @@
...
-    # TODO: Unify torch and tf config settings:
-    # Learning rate for adam optimizer for the user choice model.
-    "lr_choice_model": 1e-3,  # Only relevant for framework=torch.
-    # Learning rate for adam optimizer for the q model.
-    "lr_q_model": 1e-3,  # Only relevant for framework=torch.
-    # TODO: Unify torch and tf config settings:
-    # Learning rate for adam optimizer for the q model.
-    "lr": 0.00025,  # Only relevant for framework=tf|tf2.
+    # Learning rate for RMSprop optimizer for the q-model.
+    "lr": 0.00025,
```
Comment (avnishn) [inline rllib/agents/slateq/slateq.py:110]: this is pretty low -- is this to avoid exploding q values? Are we using target q networks for regularization in torch/tf anymore?

File: rllib/agents/slateq/slateq.py [hunk 2]
```diff
@@ -113,21 +106,17 @@
...
     # Learning rate schedule.
     # In the format of [[timestep, value], [timestep, value], ...]
     # A schedule should normally start from timestep 0.
-    "lr_schedule": None,  # Only relevant for framework=tf|tf2.
+    "lr_schedule": None,
+    # Learning rate for adam optimizer for the user choice model.
+    "lr_choice_model": 1e-3,  # Only relevant for framework=torch.
```
Comment (avnishn) [inline rllib/agents/slateq/slateq.py:116]: how come this is only relevant for torch? If its not in the TF implementation, should we drop this?

File: rllib/agents/slateq/slateq_tf_policy.py [hunk 3]
```diff
@@ -237,13 +236,13 @@ def action_distribution_fn(
...
         per_slate_q_values = get_per_slate_q_values(
-            policy.slates, slate_size, score_no_click, scores, q_values
+            policy.slates, score_no_click, scores, q_values
         )
     model.slates = policy.slates
     return per_slate_q_values, Categorical, []
 
 
-def get_per_slate_q_values(slates, slate_size, s_no_click, s, q):
+def get_per_slate_q_values(slates, s_no_click, s, q):
```
Comment (avnishn) [inline rllib/agents/slateq/slateq_tf_policy.py:245]: was the removed parameter previously unused? I can't find where it was removed.

File: rllib/agents/slateq/slateq_torch_model.py [hunk 4]
```diff
@@ -13,71 +13,64 @@
...
+    ):
         """Initializes a QValueModel instance.
 
-        Args:
-            embedding_size: The sum of user- and doc embedding sizes. This is the input
-                dimension going into the first layer of the MLP.
-            q_hiddens: List of dense layer sizes to build the MLP by.
+        Each document candidate receives one full Q-value stack, defined by
+        `fcnet_hiddens_per_candidate`. The input to each of these Q-value stacks
+        is always {[user] concat [document[i]] for i in document_candidates}.
```
Comment (avnishn) [inline rllib/agents/slateq/slateq_torch_model.py:25]: Does q-value stack mean 1 q value per candidate document? That's what I've inferred from this statement.

File: rllib/agents/slateq/slateq_torch_model.py [hunk 5]
```diff
@@ -13,71 +13,64 @@
...
-        for n in q_hiddens:
-            layers.append(nn.Linear(ins, n))
-            layers.append(nn.LeakyReLU())
-            ins = n
-        layers.append(nn.Linear(ins, 1))
-        self.layers = nn.Sequential(*layers)
+        self.orig_obs_space = obs_space
+        self.embedding_size = self.orig_obs_space["doc"]["0"].shape[0]
+        self.num_candidates = len(self.orig_obs_space["doc"])
+        assert self.orig_obs_space["user"].shape[0] == self.embedding_size
```
Comment (avnishn) [inline rllib/agents/slateq/slateq_torch_model.py:36]: should we add a message here to go with the assert? Or is this more a sanity check that the user probably won't hit.

File: rllib/agents/slateq/slateq_torch_model.py [hunk 6]
```diff
@@ -13,71 +13,64 @@
...
 
-    def forward(self, user: TensorType, doc: TensorType) -> TensorType:
-        """Evaluate the user-doc Q model
+        self.q_nets = nn.ModuleList()
+        for i in range(self.num_candidates):
+            layers = nn.Sequential()
+            ins = 2 * self.embedding_size
```
Comment (avnishn) [inline rllib/agents/slateq/slateq_torch_model.py:41]: why 2 * the embedding size?

Comment (avnishn) [reply rllib/agents/slateq/slateq_torch_model.py:41]: and not just the embedding size.

**Review by avnishn — COMMENTED** [review]
I'm a little lost as to what the difference is between the old implementation and the new implementation -- have we just greatly simplified the q function? 

I had some other random questions as well. Thanks :)
